File: addons/openerp/static/javascript/o2m.js
////////////////////////////////////////////////////////////////////////////////
//
// Copyright (C) 2007-TODAY Tiny ERP Pvt Ltd. All Rights Reserved.
//
// $Id$
//
// Developed by Tiny (http://openerp.com) and Axelor (http://axelor.com).
//
// The OpenERP web client is distributed under the "OpenERP Public License".
// It's based on Mozilla Public License Version (MPL) 1.1 with following 
// restrictions:
//
// -   All names, links and logos of Tiny, OpenERP and Axelor must be 
//     kept as in original distribution without any changes in all software 
//     screens, especially in start-up page and the software header, even if 
//     the application source code has been changed or updated or code has been 
//     added.
//
// -   All distributions of the software must keep source code with OEPL.
// 
// -   All integrations to any other software must keep source code with OEPL.
//
// If you need commercial licence to remove this kind of restriction please
// contact us.
//
// You can see the MPL licence at: http://www.mozilla.org/MPL/MPL-1.1.html
//
////////////////////////////////////////////////////////////////////////////////

var One2Many = function(name, inline) {

    this.name = name;
    this.inline = inline > 0;

    this.model = openobject.dom.get(name + '/_terp_model').value;
    this.mode = openobject.dom.get(name + '/_terp_view_type').value;

    if (openobject.dom.get(name + '/_terp_default_get_ctx'))
        this.default_get_ctx = openobject.dom.get(name + '/_terp_default_get_ctx').value;

    var parent_prefix = name.indexOf('/') > -1 ? name.slice(0, name.lastIndexOf('/') + 1) : '';

    this.parent_model = openobject.dom.get(parent_prefix + '_terp_model').value;
    this.parent_id = openobject.dom.get(parent_prefix + '_terp_id').value;
    this.parent_context = openobject.dom.get(parent_prefix + '_terp_context').value;
    this.parent_view_id = openobject.dom.get(parent_prefix + '_terp_view_id').value;

    // hide new button when editors are visible
    if (this.mode == 'tree' && this.inline) {
        var self = this;
        this.btn_new = openobject.dom.get(this.name + '_btn_');
        MochiKit.Signal.connect(ListView(this.name), 'onreload', function(evt) {
            self.btn_new.style.display = ListView(self.name).$getEditors().length > 0 ? 'none' : '';
        });
    }
};

One2Many.prototype = {

	remove: function(id) {

	if (!confirm(_('Do you really want to delete record ?'))) {
            return false;
    }

	var req = openobject.http.postJSON('/openerp/openo2m/delete', {'model': this.model, 'id': id});

	req.addCallback(function(obj) {
            if (obj.error) {
                jQuery.fancybox(obj.error, {scrolling: 'no'});
            }
            else {
                window.location.reload();
			}
    	});
	},

    create: function() {

        if (!this.parent_id || this.parent_id == 'False' || this.mode == 'form') {
            return submit_form('save', this.name);
        }

        if (this.mode == 'tree' && this.inline) {

            if (this.default_get_ctx) {
                var self = this;
                var req = eval_domain_context_request({source: this.name, context: this.default_get_ctx});
                req.addCallback(function(res) {
                    ListView(self.name).create(res.context);
                });
            } else {
                ListView(this.name).create();
            }
        } else {
            this.edit(null);
        }
    },

    edit: function(id, readonly) {
        var names = this.name.split('/');
        if (id!='False' && id==0) {
        	return error_display(_('To edit Record, please first save it.'));
        }
        var parents = [];
        // get the required view params to get proper view
        var params = {
            '_terp_view_params/_terp_model': openobject.dom.get('_terp_model').value,
            '_terp_view_params/_terp_id': openobject.dom.get('_terp_id').value,
            '_terp_view_params/_terp_ids': openobject.dom.get('_terp_ids').value,
            '_terp_view_params/_terp_view_ids': openobject.dom.get('_terp_view_ids').value,
            '_terp_view_params/_terp_view_mode': openobject.dom.get('_terp_view_mode').value,
            '_terp_view_params/_terp_context': openobject.dom.get('_terp_context').value || {},
            '_terp_view_params/_terp_view_type': 'form'
        };

        while (names.length) {
            parents.push(names.shift());
            var prefix = parents.join('/');

            params['_terp_view_params/' + prefix + '/_terp_model'] = openobject.dom.get(prefix + '/_terp_model').value;
            params['_terp_view_params/' + prefix + '/_terp_id'] = openobject.dom.get(prefix + '/_terp_id').value;
            params['_terp_view_params/' + prefix + '/_terp_ids'] = openobject.dom.get(prefix + '/_terp_ids').value;
            params['_terp_view_params/' + prefix + '/_terp_view_ids'] = openobject.dom.get(prefix + '/_terp_view_ids').value;
            params['_terp_view_params/' + prefix + '/_terp_view_mode'] = openobject.dom.get(prefix + '/_terp_view_mode').value;
            params['_terp_view_params/' + prefix + '/_terp_context'] = openobject.dom.get(prefix + '/_terp_context').value || {};
            params['_terp_view_params/' + prefix + '/_terp_view_type'] = 'form';
        }

        jQuery.extend(params, {
            _terp_parent_model: this.parent_model,
            _terp_parent_id: this.parent_id,
            _terp_parent_view_id: this.parent_view_id,
            _terp_o2m: this.name,
            _terp_o2m_model: this.model,
            _terp_editable: readonly ? 0 : 1
        });

        if(id != null) {
            params['_terp_o2m_id'] = id;
        }
        if (id && id != 'False' && !this.default_get_ctx) {
            jQuery.o2m(params);
            return;
        }

        eval_domain_context_request({
            source: this.name,
            context : this.default_get_ctx
        }).addCallback(function(res) {
            jQuery.o2m(jQuery.extend(params, {
                _terp_o2m_context: res.context,
                _terp_parent_context: this.parent_context
            }));
        });
    },

    setReadonly: function(readonly) {
        var btn=MochiKit.DOM.getElement(this.name+'_btn_');
        var grid=MochiKit.DOM.getElement(this.name+'_grid');
        var edit=MochiKit.DOM.getElement(this.name + '/_terp_editable');
        var rows = jQuery('table[id='+this.name+'_grid] tr.grid-row');
        if (readonly) {
            jQuery('table.one2many[id$="'+this.name+'"]').addClass('m2o_readonly');
            if(btn){btn.style.display='none';}
            MochiKit.Base.map(function (el) {el.style.display='none'},MochiKit.Selector.findChildElements(grid,['.selector']));
            edit.value= 0;
            if (rows && rows.length) {
                rows.each(function(index, row) {
                    jQuery(row).unbind('click');});
            }
        }
        else{
            if(btn){btn.style.display='';}
            MochiKit.Base.map(function (el) {el.style.display=''},MochiKit.Selector.findChildElements(grid,['.selector']));
            edit.value = 1;
            if (rows && rows.length) {
                rows.each(function(index, row) {
                    jQuery(row).bind('click');});
            }
        }
    }
};

(function ($) {
    /**
     * Frame counter, used to uniquify (monotonously increasing) frame id in
     * order to ensure we avoid collisions between o2m frames (in case we
     * have nested o2ms somehow), as we need to get a frame we can target
     * with a form submission.
     */
    var frame_counter = 0;

    function frame_data($this, data) {
        return $($this.attr('frameElement')).data(data);
    }

    /**
     * Opens an o2m dialog linked to the provided <code>$this</code> window,
     * with the selected options.
     *
     * @param $this the parent window of the opened dialog, contains the
     * input to fill with the selected o2m value if any
     * @param options A map of options to provide to the newly opened iframe
     * call.
     */
    function open($this, options) {
        var frame_identifier = 'test-frame' + frame_counter++;
        var $frame = $('<iframe>', {
            src: 'about:blank',
            // never sure whether the iframe is targetted by name or by id,
            // so let's just set both
            id: frame_identifier,
            name: frame_identifier,
            frameborder: 0
        }).data('source-window', $this[0])
          .data('list', options['_terp_o2m'])
          .appendTo(document.documentElement)
          .dialog({
              modal: true,
              width: 640,
              height: 480,
              close: function () {
                  jQuery(this).dialog('destroy').remove();
              }
          });
        var $form = jQuery('<form>', {
            method: 'POST',
            action: '/openerp/openo2m/edit',
            target: frame_identifier
        });
        jQuery.each(options, function (key, value) {
            $form.append(jQuery('<input>', {
                type: 'hidden',
                name: key,
                value: value
            }));
        });
        $form.submit();
        return $frame;
    }


    var fetched_attrs = ['id', 'name', 'value', 'kind', 'class', 'domain',
                         'context', 'relation'];

    
    function initialize($frame_window) {
        // Within initialize, <code>jQ</code> is the "toplevel" jQuery whereas
        // <code>$</code> is the jQuery from within the iframe
        var jQ = jQuery;
        var $ = $frame_window[0].jQuery;

        var $form = $('#view_form');

        var list_id = frame_data($frame_window, 'list');
        jQ('#view_form :input[name]:not(:button)').each(function () {
            if (this.name.indexOf(list_id) == 0) {
                return;
            }
            var $this = $(this);
            var attrs = {};
            $.each(fetched_attrs, function (index, attribute) {
                var attr = $this.attr(attribute);
                if(!attr) { return; }
                attrs[attribute] = attr;
            });
            $form.append($('<input>', $.extend(attrs, {
                type: 'hidden',
                disabled: 'disabled',
                value: $this.val()
            })));
        });

        var lc = parseInt($('#_terp_load_counter').val(), 10) || 0;
        if(lc) {
            $.o2m('refresh');
        }
    }

    /**
     * Refreshes the backing ListView to make the newly added items appear.
     *
     * @param $this the window of the dialog to close
     */
    function refresh($this) {
        setTimeout(function () {
            frame_data($this, 'source-window')
                .ListView(frame_data($this, 'list'))
                    .reload(null, 1);
        })
    }

    /**
     * Closes the o2m dialog it was called from (represented by
     * <code>$this</code>.
     *
     * @param $this the window of the dialog to close
     */
    function close($this) {
        $($this.attr('frameElement')).dialog('close');
        return null;
    }

    /**
     * Manage o2m dialogs for this scope
     * <ul>
     *  <li><p>Called with only options, opens a new o2m dialog linking to the
     *         current scope.</p></li>
     *  <li><p>Called with the <code>"init"</code> command, performs the
     *         initial setup for the o2m dialog within the opened iframe
     *  </p></li>
     *  <li><p>Called with the <code>"close"</code> command, closes the o2m
     *         dialog it was invoked from and focuses its parent scope.
     *  </p></li>
     *  <li><p>Called with the <code>"refresh"</code> command, reloads
     *         the backing ListView to make the new items in it appear.
     *  </p></li>
     * </ul>
     *
     * @returns the o2m container (iframe) if one was created
     */
    $.o2m = function () {
        // $this should be the holder for the window from which $.o2m was
        // originally called, even if $.o2m() was bubbled to the top of
        // the window stack.
        var $this;
        if(this == $) $this = $(window);
        else $this = $(this);
        if(window != window.top) {
            return window.top.jQuery.o2m.apply($this[0], arguments);
        }
        // We're at the top-level window, $this is the window from which the
        // original $.o2m call was performed, window being the current window
        // level.
        switch(arguments[0]) {
            case 'close':
                return close($this);
            case 'refresh':
                return refresh($this);
            case 'init':
                return initialize($this);
            default:
                return open($this, arguments[0]);
        }
    };
})(jQuery);

File: addons/openerp/widgets/form/_o2m.py
###############################################################################
#
#  Copyright (C) 2007-TODAY OpenERP SA. All Rights Reserved.
#
#  $Id$
#
#  Developed by OpenERP (http://openerp.com) and Axelor (http://axelor.com).
#
#  The OpenERP web client is distributed under the "OpenERP Public License".
#  It's based on Mozilla Public License Version (MPL) 1.1 with following 
#  restrictions:
#
#  -   All names, links and logos of OpenERP must be kept as in original
#      distribution without any changes in all software screens, especially
#      in start-up page and the software header, even if the application
#      source code has been changed or updated or code has been added.
#
#  You can see the MPL licence at: http://www.mozilla.org/MPL/MPL-1.1.html
#
###############################################################################
import time
import copy
import cherrypy

from openerp.utils import TinyDict, expr_eval, rpc
from openerp.widgets import TinyInputWidget, register_widget
from openerp.widgets.screen import Screen

__all__ = ["O2M", "OneToMany"]

#O2M tuple format:
#
#(0, _, {values}): CREATE new record (linked to current) with values $values
#(1, id, {values}): UPDATE linked record $id with values $values
#(2, id, _): DELETE linked record $id
#(3, id, _): FORGET linked record $id (removes relation, but not linked record)
#(4, id, _): LINK TO existing record $id
#(5, _, _): FORGET ALL linked records
#(6, _, [ids]): REPLACE LINKS to links to $ids
class OneToMany(object):
    @classmethod
    def create(cls, values=None, **kw):
        """ Returns the command used to create a new o2m linked to the
        current record.

        The values for the record to create can be provided as a dict (or an
        iterable of pairs) and/or as **kwargs, as for the `dict` constructor
        or `dict.update`

        :param values: a dictionary or pair iterable
        :type values: {str: value}
        :rtype: (int, int, dict)
        """
        return 0, False, dict(values or {}, **kw)
    @classmethod
    def replace_all(cls, *with_ids):
        """ Returns the command used to replace all values of an o2m with the
         provided ids. Equivalent to a FORGET ALL followed by a LINK TO

         :param with_ids: the ids to set in the o2m
         :type with_ids: int...
         :rtype: (int, ?, [int])
        """
        return 6, False, with_ids
    @classmethod
    def update(cls, record_id, values):
        """ Returns the command used to update an existing o2m record with the
         provided values

         :param record_id: the id of the o2m record to update
         :type record_id: int
         :param values: the new values to set on the record
         :type values: dict
         :rtype: (int, int, dict)
        """
        return 1, record_id, values

class O2M(TinyInputWidget):
    """One2Many widget
    """
    template = "/openerp/widgets/form/templates/one2many.mako"
    params = ['id', 'parent_id', 'new_attrs', 'pager_info', 'switch_to',
              'default_get_ctx', 'source', 'view_type', 'default_value',
              'edition']
    member_widgets = ['screen']

    form = None
    valign = "top"
    default_value = []

    def __init__(self, **attrs):
        #FIXME: validation error in `Pricelist Version`
        attrs['required'] = False

        super(O2M, self).__init__(**attrs)

        self.new_attrs = { 'text': _("New"), 'help': _('Create new record.')}
        self.default_get_ctx = attrs.get('default_get', {}) or attrs.get('context', {})

        # get top params dictionary
        params = cherrypy.request.terp_params
        self.source = params.source
        self.edition = params.o2m_edit
        pprefix = ''
        if '/' in self.name:
            pprefix = self.name[:self.name.rindex('/')]

        pparams = params.chain_get(pprefix)
        if (pparams and not pparams.id) or (not pparams and not params.id):
            self.new_attrs = { 'text': _("Save/New"), 'help': _('Save parent record.')}

        self.parent_id = params.id
        if pparams:
            self.parent_id = pparams.id

        # get params for this field
        current = params.chain_get(self.name)

        self.model = attrs['relation']
        self.link = attrs.get('link', '')
        self.onchange = None # override onchange in js code

        view = attrs.get('views', {})
        mode = str(attrs.get('mode', 'tree,form')).split(',')

        self.view = view

        view_mode = mode
        view_type = mode[0]
        self.view_type = view_type
        
        if not current:
            current = TinyDict()

        if current.view_mode: view_mode = current.view_mode
        if current.view_type: view_type = current.view_type

        self.switch_to = view_mode[-1]
        if view_type == view_mode[-1]: self.switch_to = view_mode[0]

        ids = attrs.get('value') or []
        if not isinstance(ids, list):
            ids = [ids]

        if ids:
            if isinstance(ids[0], dict):
                current.default_data = ids
                for item in current.default_data:
                    self.default_value.append(
                        OneToMany.create(item))
                    item['id'] = 0
                ids = []
            elif isinstance(ids[0], tuple):
                [current_id[1] for current_id in ids]
        
        id = (ids or None) and ids[0]
        
        if self.name == self.source or self.name == params.source:
            if params.sort_key and ids:
                domain = current.domain or []
                domain.append(('id', 'in', ids))
                ids = rpc.RPCProxy(self.model).search(domain, current.offset, current.limit, params.sort_key + ' '+params.sort_order, current.context)
                id = ids[0]
        if current and params.source and self.name in params.source.split('/'):
            id = current.id

        id = id or None
                
        current.model = self.model
        current.id = id
        current.ids = ids
        current.view_mode = view_mode
        current.view_type = view_type
        current.domain = current.domain or []
        current.context = current.context or {}

        group_by_ctx = ''
        if self.default_get_ctx:
            ctx = dict(cherrypy.request.terp_record,
                       context=current.context,
                       active_id=self.parent_id or False)
            ctx[attrs['name']] = ids
            # XXX: parent record for O2M
            #if self.parent:
            #    ctx['parent'] = EvalEnvironment(self.parent)

            try:
                context = ctx.copy()
                ctx = expr_eval("dict(%s)" % self.default_get_ctx, context)
                ctx.update(expr_eval("dict(%s)" % attrs.get('context', '{}'), context))
                current.context.update(ctx)
            except:
                pass

            if ctx and ctx.get('group_by'):
                group_by_ctx = ctx.get('group_by')

        current.offset = current.offset or 0
        current.limit = current.limit or 50
        current.count = len(ids or [])

        # Group By for one2many list.
        if group_by_ctx:
            current.group_by_ctx = group_by_ctx
            current.domain = [('id', 'in', ids)]

        if current.view_type == 'tree' and self.readonly:
            self.editable = False

        if 'default_name' in current.context:
            del current.context['default_name']

        if self.view_type == 'tree' and pparams:
            self.editable = bool(pparams.id)

        self.screen = Screen(current, prefix=self.name, views_preloaded=view,
                             editable=self.editable, readonly=self.readonly,
                             selectable=0, nolinks=self.link, _o2m=1)
        
        self.id = id
        self.ids = ids

        if view_type == 'tree':
            self.id = None

        elif view_type == 'form':
            records_count = len(self.screen.ids or [])

            current_record = 0
            if records_count and self.screen.id in self.screen.ids:
                current_record = self.screen.ids.index(self.screen.id) + 1
                self.pager_info = _('%d of %d') % (current_record, records_count)
            else:
                self.pager_info = _('- of %d') % (records_count)

    def get_value(self):

        if not self.ids:
            return []

        values = getattr(self.screen.widget, 'values', [])

        return [(1, val.get('id', False), val) for val in values]

register_widget(O2M, ["one2many", "one2many_form", "one2many_list"])
